sqp/base.py

```diff
@@ -1,172 +1,379 @@
+
+    Returns:
+        Value of merit function
+    """
     return f_x + mu * c_norm
 
-# should be done (D_1(phi_1) function in book)
+
 def l1_merit_directional_gradient(
         f_grad: np.ndarray, 
         p: np.ndarray, 
         mu: float, 
-        c_norm: float):
+        c_norm: float) -> float:
+    """
+    Calculates the directional gradient of the merit function wrt p at x.
+
+    Args:
+        f_grad: Derivative of function at current iterate x
+        p: Direction found with QP
+        mu: Mu that fulfills 18.36
+        c_norm: L1-norm of constraint values at current iterate x
+
+    Returns:
+        Directional gradient of merit function wrt p
+    """
     return f_grad @ p - mu * c_norm
 
 
 def find_lambda_of_qp(
         problem: QuadraticProblem,
         x_solution: np.ndarray) -> np.ndarray:
+    """
+    Computes the lambda vector (Lagrange multipliers) for a given solution x and a quadratic sub-problem
+
+    Args:
+        problem: Quadratic sub-problem
+        x_solution: Solution x of KKT-system for quadratic sub-problem
 
+    Returns:
+        Lambda vector (Lagrange multipliers) for given solution x
+    """
+
+    # start with 0 for all constraints
+    # s.t. all non-active constraints i already have the correct lambda_i=0
     lambda_ = np.zeros(len(problem.constraints))
 
-    g = problem.G @ x_solution + problem.c
+    # find all active constraints at the current point
     active_constraint_mask = np.array([c.as_equality().holds(x_solution) for c in problem.constraints])
     if np.all(~active_constraint_mask):
-        # no active constraints, all lambda=0
+        # no active constraints, early exit with lambda=0
         return lambda_
 
+    # compute derivative of quadratic function
+    g = problem.G @ x_solution + problem.c
+
+    # calculate constraint jacobian for active constraints
+    # note: slicing will work as constraints are always numpy arrays
+    # noinspection PyTypeChecker
     A, _ = combine_linear([c.c for c in problem.constraints[active_constraint_mask]])
+
+    # solve linear system Ax=g which effectively ensures KKT-condition 12.34a
     lambda_[active_constraint_mask] = np.linalg.lstsq(A.T, g)[0]
 
     return lambda_
 
-# should be done
+
 def transform_sqp_to_linear_constraint(
         constraint: Constraint, 
         c_i: float, 
-        c_i_grad: np.ndarray):
+        c_i_grad: np.ndarray) -> LinearConstraint:
+    """
+    Creates a linear constraint for a quadratic sub-problem based on a non-linear constraint (see 18.11).
+
+    Args:
+        constraint: Original non-linear constraint
+        c_i: Value of constraint at x
+        c_i_grad: Value of derivative of constraint at x
+
+    Returns:
+        Linear constraint for quadratic sub-problem
+    """
 
     return LinearConstraint(
         c=LinearCallable(a=c_i_grad, b=-c_i),
         equation_type=constraint.equation_type)
 
-# should be done
+
 def create_iterate_quadratic_problem(
```

Comment:
Das is auch wieder iwie was, was iMo entweder in des `MinimizationProblem` oder ins `QuadraticProblem` gehört, weils recht viele parameter hat, die in den Klassen scho gespeichert sind. Des hält wieder des Interface klein und den Code kompakter (imo).

Reply:
imo willst du die datenklassen zu sehr mit den algorithmen vermischen^^
es gibt scho an grund warum de in shared oder in dem fall quadratic liegen und ned in sqp usw^^

was ginge, wäre a klasse SqpIterateQuadraticProblem(QuadraticProblem) z.b., welches im folder sqp lebt

Reply:
Ich versteh schon was du meinst, aber es is halt wieder einfach a Konstruktor, der abgesehen vielleicht von `B`, alle parameter von derer Funktion direkt mit reference zu `self` sich selber herzaubern könnte.

Aber ja, ma kann argumentieren, dass da a Funktion auch sinnvoll wär weil ma eh schon die ganzen Sachen bei der Anwendung drüber berechnen.